Sources/Utilities/ImageTools/ImageTools.swift

```diff
@@ -0,0 +1,358 @@
+    case invalidPercentage
+    case compression
+    case invalidSize
+  }
+  
+  enum ImageFormat {
+    case jpeg(compressionRatio: CGFloat)
+    case png
+  }
+  
+  /// Saves the given image to the photo library asynchronously.
+  ///
+  /// This function saves the provided UIImage to the user's photo library. It performs
+  /// the operation asynchronously and throws an error if the save operation fails.
+  ///
+  /// - Parameters:
+  ///   - image: The UIImage to be saved to the photo library.
+  /// - Example:
+  /// ```swift
+  /// do {
+  ///   let image = UIImage(named: "exampleImage")!
+  ///   try await saveImageToPhotoLibrary(image)
+  ///   print("Image saved successfully.")
+  /// } catch {
+  ///   print("Failed to save image: \(error)")
+  /// }
+  /// ```
+  /// - Note: This function should be called from an asynchronous context using `await`.
+  /// - Throws: An error if the save operation fails.
+  func saveToPhotoLibrary(image: UIImage) async throws {
+    try await withCheckedThrowingContinuation { continuation in
+      let continuationWrapper = ContinuationWrapper(continuation: continuation)
+      UIImageWriteToSavedPhotosAlbum(
+        image,
+        self,
+        #selector(saveCompleted),
+        Unmanaged.passRetained(continuationWrapper).toOpaque()
+      )
+    }
+  }
+  
+  /// Downsizes the given image to the specified target size, asynchronously, respecting aspect ratio.
+  /// - Parameters:
+  ///   - image: The original UIImage to be downsized.
+  ///   - targetSize: The desired size to which the image should be downsized.
+  /// - Returns: An optional UIImage that is the result of the downsizing operation.
+  /// - Example:
+  /// ```swift
+  /// do {
+  ///   let image = UIImage(named: "exampleImage")!
+  ///   let targetSize = CGSize(width: 100, height: 100)
+  ///   let resizedImage = await downsize(image, to: targetSize)
+  ///   imageView.image = resizedImage
+  /// } catch {
+  ///   print("Failed to downsize image: \(error)")
+  /// }
+  /// ```
+  /// - Note: This function should be called from an asynchronous context using `await`.
+  func downsize(image: UIImage, toTargetSize targetSize: CGSize) async throws -> UIImage {
+    guard !targetSize.width.isZero, !targetSize.height.isZero else {
+      throw ImageError.invalidSize
+    }
+    
+    return await Task.detached(priority: .high) {
+      let originalSize = image.size
+      let widthRatio = targetSize.width / originalSize.width
+      let heightRatio = targetSize.height / originalSize.height
+      let scaleFactor = min(widthRatio, heightRatio)
+      let newSize = CGSize(
+        width: floor(originalSize.width * scaleFactor),
+        height: floor(originalSize.height * scaleFactor)
+      )
+      
+      let renderer = UIGraphicsImageRenderer(size: newSize)
+      let newImage = renderer.image { _ in
+        image.draw(in: CGRect(origin: .zero, size: newSize))
+      }
+      
+      return newImage
+    }.value
```

Comment:
I do not like making assumptions about how the user will use the code. Why wrap the synchronous code into an async block, especially with a detached task? 
The concurrency overhead might be higher than the actual image rendering.

Reply:
Good thought... not that you mention it, I never to async in image resizing ... 🤔

Reply:
Image processing tasks are rather expensive and will block the main thread if not used correctly. Which we want to avoid at all cost.

I agree with you to a point, that the call site should handle that. That's the go-to solution for many apis.

In most cases, we are processing larger images that do take portion of the time to process and blocks the main thread. Using this api, dev doesn't need to think how to properly use it not to block the UI.

Similar to how Kingfisher works. You don't actually care how it does it, you just need the result without sacrificing user experience.

We can go the other direction and have all this apis sync, leave that decision to the caller and don't care about the result. Which will in most cases result to a bad ux. We all know we tend to be lazy, forget about the performance when we are mostly testing on the latest, most performant hardware. And we only realise the issue when the customers reports laggy experience with older devices. 
On the other hand by "enforcing" user async context, it will never lead to such situation.

Reply:
Are you sure `Task.detached` is the solution? As far as I can tell, the documentation does not mention on which thread the newly created task would execute. For all you know, the code can be executed on the main thread.

Reply:
You are correct. You cannot define the thread on which Task will run. You can instead define the priority only. And the Task itself will figure out what's the best thread to run on based on the available system resources. We never actually care about the thread, but the queues (concurrency runs on top of GCD).

One thing I need to change it to remove the detached task. It's not needed and in many cases less safe than regular Task.